### DISCUSSION TITLE
Testing with API Client node module

### DISCUSSION THREAD

**Discussion by rvalle** [artilleryio/artillery#1202] [Artillery]
Hi!

I wonder if it is possible to test neither with HTTP or WS but using a node module that implements an API client.

Our API client is very complex, with persistent connections, pub/sub, and loads of cryptographic functions as part of the request/reply process.

Testing with RAW http/ws will not allow me to generate a load pattern that in any way resembles actual load of the service.

I am new to artillery, but I see it is also build on nodejs, and I wonder if I can create some kind of plugin, that will execute the actually network requests and it is implemented with own API client.

Can this be done with artillery?

**Comment by hassy** [comment]
hi @rvalle! it sounds like writing a custom engine is the way forward. A custom engine would allow you to expose actions via YAML, and map them onto operations in the API client as needed (that's what the built-in engines for HTTP/WS/Socket.io do). You'd have full control over what happens underneath in terms of connections, data flowing back and forth from the API etc.

There's a couple of engines you could look at to get started, e.g.:

https://github.com/artilleryio/artillery-engine-kinesis

https://github.com/artilleryio/artillery-engine-kinesis

(and of course the built-in engines, of which the HTTP one is the most complex)
